cardano-crypto-class/src/Cardano/Crypto/DSIGN/EcdsaSecp256k1.hs

```diff
@@ -55,9 +81,198 @@ import Cardano.Crypto.DSIGN.Class (
+              _ -> Right ()
+    genKeyDSIGN seed = runMonadRandomWithSeed seed $ do
+      bs <- getRandomBytes 32
+      case psbFromByteStringCheck bs of 
+        Nothing -> error "genKeyDSIGN: Failed to generate key unexpectedly"
+        Just psb -> pure $ SignKeyEcdsaSecp256k1 psb
+    {-# NOINLINE rawSerialiseSigDSIGN #-}
+    rawSerialiseSigDSIGN (SigEcdsaSecp256k1 psb) = 
+      psbToByteString @SECP256K1_ECDSA_SIGNATURE_BYTES . unsafeDupablePerformIO $ do
+        psbUseAsSizedPtr psb $ \psp -> 
+          psbCreateSized $ \dstp -> 
+            void $ secpEcdsaSignatureSerializeCompact secpCtxPtr dstp psp
+    {-# NOINLINE rawSerialiseVerKeyDSIGN #-}
+    rawSerialiseVerKeyDSIGN (VerKeyEcdsaSecp256k1 psb) = 
+      psbToByteString . unsafeDupablePerformIO . psbUseAsSizedPtr psb $ \psp -> 
+        psbCreateSized @SECP256K1_ECDSA_PUBKEY_BYTES $ \(SizedPtr ptr) -> do
+          let dstp = castPtr ptr
+          let len :: CSize = fromIntegral . natVal $ Proxy @SECP256K1_ECDSA_PUBKEY_BYTES
+          -- This is necessary because of how the C API handles checking writes:
+          -- maximum permissible length is given as a pointer, which is
+          -- overwritten to indicate the number of bytes we actually wrote; if
+          -- we get a mismatch, then the serialization failed. While an odd
+          -- choice, we have to go with it.
+          alloca $ \(lenPtr :: Ptr CSize) -> do
+            poke lenPtr len
+            void $ secpEcPubkeySerialize secpCtxPtr dstp lenPtr psp secpEcCompressed
+            writtenLen <- peek lenPtr
+            unless (writtenLen == len) (error "rawSerializeVerKeyDSIGN: Did not write correct length")
+    rawSerialiseSignKeyDSIGN (SignKeyEcdsaSecp256k1 psb) = psbToByteString psb
+    {-# NOINLINE rawDeserialiseSigDSIGN #-}
+    rawDeserialiseSigDSIGN bs = 
+      SigEcdsaSecp256k1 <$> (psbFromByteStringCheck bs >>= go)
+      where
+        go :: 
+          PinnedSizedBytes SECP256K1_ECDSA_SIGNATURE_BYTES -> 
+          Maybe (PinnedSizedBytes SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL)
+        go psb = unsafeDupablePerformIO . psbUseAsSizedPtr psb $ \psp -> do
+          -- This is a slightly odd order of operations, but it's forced on us,
+          -- as psbCreateSized only allows an initialization that returns the
+          -- trivial value. Instead, we allocate (but do not initialize) a
+          -- PinnedSizedBytes of the right size, then try using it. This is
+          -- safe, as if deserialization fails, then we never get to see the
+          -- PinnedSizedBytes this would produce (ostensibly full of junk), and
+          -- if deserialization succeeds, it'll overwrite the contents anyway.
+          sigPsb <- psbCreateSized (\_ -> pure ())
+          res <- psbUseAsSizedPtr sigPsb $ \sigp -> 
+                  secpEcdsaSignatureParseCompact secpCtxPtr sigp psp
+          pure $ case res of 
+            1 -> pure sigPsb
+            _ -> Nothing
```

Comment:
I am not against the approach that you take here, it does seem safe, I just think it would be more useful to add functions that can handle this sort of pattern:

```haskell
-- | Make sure to not return the pointer as the result, this would be very unsafe.
psbCreateResult :: forall n. KnownNat n => (Ptr Word8 -> IO r) -> IO (PinnedSizedBytes n, r)
psbCreateResult k = do
    let size :: Int
        size = fromInteger (natVal (Proxy :: Proxy n))
    mba <- newPinnedByteArray size
    res <- k (mutableByteArrayContents mba)
    arr <- unsafeFreezeByteArray mba
    return (PSB arr, res)

-- | Make sure to not return the pointer as the result, this would be very unsafe.
psbCreateSizedResult :: forall n. KnownNat n => (SizedPtr n -> IO r) -> IO (PinnedSizedBytes n, r)
psbCreateSizedResult k = psbCreateResult (k . SizedPtr . castPtr)

psbCreate :: forall n. KnownNat n => (Ptr Word8 -> IO ()) -> IO (PinnedSizedBytes n)
psbCreate = fmap fst . psbCreateResult

psbCreateSized :: forall n. KnownNat n => (SizedPtr n -> IO ()) -> IO (PinnedSizedBytes n)
psbCreateSized k = fmap fst . psbCreateSizedResult
```

Reply:
I agree with having such functions available, but it was definitely outside of the scope of my work here. If @michaelpj agrees, I'm happy to add this functionality (and use it).

Reply:
Yes, please do. I think that factoring the code well so as to ensure the ongoing maintainability of the codebase is _always_ in scope :)